R/GeospatialFunctions.R

```diff
@@ -0,0 +1,611 @@
+    
+    # bounding box (with some minor wiggle) of user's WQP data
+    suppressMessages(suppressWarnings({
+      
+      bbox <- data %>% 
+        sf::st_bbox(.) %>%
+        # convert bounding box to characters
+        toString(.) %>% 
+        # encode for use within the API URL
+        urltools::url_encode(.)
+      
+    }))
+    
+    # The ATTAINS API has a limit of 2000 features that can be pulled in at once.
+    # Therefore, we must split the call into manageable "chunks" using a moving
+    # window of what features to pull in, then munging all the separate API calls 
+    # together.
+    
+    repeat {
+      
+      query <- urltools::param_set(baseurls, key = "geometry", value = bbox) %>%
+        urltools::param_set(key = "inSR", value = our_epsg) %>%
+        # Total of 2000 features at a time...
+        urltools::param_set(key = "resultRecordCount", value = 2000) %>%
+        # ... starting at the "offset":
+        urltools::param_set(key = "resultOffset", value = offset) %>%
+        urltools::param_set(key = "spatialRel", value = "esriSpatialRelIntersects") %>%
+        urltools::param_set(key = "f", value = "geojson") %>%
+        urltools::param_set(key = "outFields", value = "*") %>%
+        urltools::param_set(key = "geometryType", value = "esriGeometryEnvelope") %>%
+        urltools::param_set(key = "returnGeometry", value = "true") %>%
+        urltools::param_set(key = "returnTrueCurves", value = "false") %>%
+        urltools::param_set(key = "returnIdsOnly", value = "false") %>%
+        urltools::param_set(key = "returnCountOnly", value = "false") %>%
+        urltools::param_set(key = "returnZ", value = "false") %>%
+        urltools::param_set(key = "returnM", value = "false") %>%
+        urltools::param_set(key = "returnDistinctValues", value = "false") %>%
+        urltools::param_set(key = "returnExtentOnly", value = "false") %>%
+        urltools::param_set(key = "featureEncoding", value = "esriDefault")
+      
+      # Fetch features within the offset window and append to list:
+      features <- suppressMessages(suppressWarnings({tryCatch({
+        geojsonsf::geojson_sf(query)
+      }, error = function(e) {
+        NULL
+      })}))
+      
+      
+      # Exit loop if no more features or error occurred
+      if (is.null(features) || nrow(features) == 0) {
+        break
+      }
+      
+      all_features <- c(all_features, list(features))
+      # once done, change offset by 2000 features:
+      offset <- offset + 2000
+      
+      if(offset == 4000){print("Your TADA data covers a large spatial range. The ATTAINS pull may take a while.")}
+      
+    }
+    
+    all_features <- dplyr::bind_rows(all_features)
+    
+  }
+  
+  final_features <- baseurls %>%
+    purrr::map(~ feature_downloader(.))
+  
+  names(final_features) <- c("ATTAINS_catchments", "ATTAINS_points", "ATTAINS_lines", "ATTAINS_polygons")
+  
+  return(final_features)
+  
+}
+
+
+#' TADA_GetATTAINS
```

Comment:
I am getting some inconsistent results when using TADA_GetATTAINS. With small TADA data frames over a smaller area (like a HUC), all the expected geometry (catchment outlines, polyons, lines, points) is returned and appears on the map with TADA_ViewATTAINS. For larger data sets across a larger area, sometimes one or more of the ATTAINS objects expected are returned with a NULL value so then don't show up with TADA_ViewATTAINS.

This is an example of a data set where I saw those issues:
test1 <- TADA_DataRetrieval(startDate = "2018-05-01", 
                           endDate = "2018-09-30", 
                           statecode = "IL", 
                           applyautoclean = TRUE)

There should be both lines and polygons, but only polygons are successfully included in the output.

Sometimes when I run the above example (or other similar ones) in fetchATTAINS, I will get an error message with a URL. At the URL, the final lines visible indicate: "exceededTransferLimit": true. 
![image](https://github.com/USEPA/TADA/assets/152432687/1bae5215-aa48-4c52-8130-3492b1120e91)

Reply:
@kathryn-willi is this something you and you team have time to look into before we merge this PR in? Or would you prefer we create a separate issue for this and address it in a separate PR later?

Reply:
Hi Cristina! We have been working on fixing the issue, but we haven't been able to reproduce the error. 🤔 I am happy to continue troubleshooting ahead of merging this PR, but I also do not want this to slow down progress on your other work. If you are comfortable with merging as-is and us creating an issue, we will continue to work towards addressing it!

Reply:
@hillarymarler could you share a more detailed reprex to help Katie reproduce the errors you ran into/check if you are still running into them now? I wonder if the ATTAINS API time out issues are intermittent (might not be reproducible), similarly to the WQP.